### PR TITLE
Add icon uploader in admin navigation links

### PR DESCRIPTION
**Pull request by joaoGabriel55**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Was added an SVG Image Picker to navigation link form. There, the user can upload svg icons (Forem recommends [Twemoji](https://twemoji.twitter.com/)) of an easy way.

*_I think that icon preview design could be improved. I am open to suggestions._ :smile: 

## Related Tickets & Documents

#11486

## QA Instructions, Screenshots, Recordings

Just go to `/admin/navigation_links` and do add or edit a Navigation Link.

### Screenshots:

Add Navigation Link

image

Edit Link

image

## Added tests?

- [ ] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [x] I need help with writing tests to upload image feature

## Added to documentation?

- [ ] [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [ ] README
- [x] No documentation needed

alt_text

### CODE REVIEW

**Comment by benhalpern** [comment]
Thanks for this! There's an internal draft RFC related to this functionality, and this PR _may_ just be perfect for that, but I think our team will wait to make a choice on this until our next planning cycle. _Just in case_ we go in a slightly different direction.

**Comment by joaoGabriel55** [comment]
Hello @benhalpern, thanks a lot for your feedback. I am glad to help your team.

**Comment by msarit** [comment]
@benhalpern @mstruve @vaidehijoshi

Hey team, per [Ben's comment](https://github.com/forem/forem/pull/12388#issuecomment-765724803) earlier, was this PR (or [associated issue](https://github.com/forem/forem/issues/11486)) picked up for our current cycle? I checked the Current Cycle Work board and did not find any related entries. Please advise 🙏🏾

**Comment by mstruve** [comment]
Thanks for the flag @msarit. We are currently comparing this to the feature proposed in the RFC and if it matches with our vision we will roll with it and close the RFC on our end.  

@joaoGabriel55 thank you so much for your patience with this. We have been implementing a new system for features and there are still loose ends we need to clean up. Thanks for hanging in there while we get it all sorted out! We really appreciate it! Also welcome to the Forem community :)

File: app/views/admin/navigation_links/_form.html.erb [hunk 1]
```diff
@@ -8,10 +8,21 @@
...
   <div class="alert alert-info">A full (external) or relative (internal) URL for the link</div>
 </div>
-<div class="form-group">
-  <%= form.label :icon %>
-  <%= form.text_area :icon, class: "form-control" %>
-  <div class="alert alert-info">An SVG icon should be pasted in the textarea</div>
+<div class="form-group" data-controller="svg-icon-upload">
+  <div nav-link-id="<%= form.object[:id] %>">
+    <%= form.label :icon %>
+    <div id="svg-icon-preview" class="<%= form.object[:id].nil? ? "" : "pb-3" %>" style="width: 24px;">
```
Comment (rhymes) [inline app/views/admin/navigation_links/_form.html.erb:14]: If it's posible, can we move the width of the preview in the CSS? BTW 24px is really small, I'd double it to 48px

File: app/javascript/admin/controllers/svg_icon_upload_controller.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    reader.onload = (content) => {
+      const result = content.target.result;
+      svgIconContent.value = result;
+      this.setSvgIconPreview(currentNavigationLinkModal, result);
+    };
+  }
+
+  setSvgIconPreview(document, content) {
+    const iconPreview = document.querySelector('#svg-icon-preview');
```
Comment (rhymes) [inline app/javascript/admin/controllers/svg_icon_upload_controller.js:42]: this `querySelector` is not needed if you use Stimulus targets

File: app/javascript/admin/controllers/svg_icon_upload_controller.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  selectSvgIcon(event) {
+    const currentNavigationLinkModal = this.targets.scope.data.element.querySelector(
+      'div',
+    );
+
+    this.clearInvalidIconTypeMessage(currentNavigationLinkModal);
+
+    const svgIconContent = currentNavigationLinkModal.querySelector(
+      '#svg-icon-content',
+    );
```
Comment (rhymes) [inline app/javascript/admin/controllers/svg_icon_upload_controller.js:13]: this `querySelector` is not needed if you use Stimulus targets, see above

File: app/javascript/admin/controllers/svg_icon_upload_controller.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  setSvgIconPreview(document, content) {
+    const iconPreview = document.querySelector('#svg-icon-preview');
+    iconPreview.innerHTML = content;
+    if (content) iconPreview.classList.add('pb-3');
+    else if (iconPreview.classList.length !== 0)
+      iconPreview.classList.remove('pb-3');
+  }
+
+  clearInvalidIconTypeMessage(document) {
+    const alertMessage = document.querySelector('#svg-icon-message-validate');
```
Comment (rhymes) [inline app/javascript/admin/controllers/svg_icon_upload_controller.js:50]: same thing about not needing to `querySelector` an element

File: app/javascript/admin/controllers/svg_icon_upload_controller.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    if (alertMessage.classList.length !== 0) {
+      alertMessage.classList.remove('alert', 'alert-danger');
+      alertMessage.innerHTML = null;
+    }
+  }
+
+  invalidIconTypeMessage(document, type) {
+    const alertMessage = document.querySelector('#svg-icon-message-validate');
```
Comment (rhymes) [inline app/javascript/admin/controllers/svg_icon_upload_controller.js:58]: same thing again

File: app/views/admin/navigation_links/_form.html.erb [hunk 6]
```diff
@@ -8,10 +8,21 @@
   <%= form.text_field :url, class: "form-control" %>
   <div class="alert alert-info">A full (external) or relative (internal) URL for the link</div>
 </div>
-<div class="form-group">
-  <%= form.label :icon %>
-  <%= form.text_area :icon, class: "form-control" %>
-  <div class="alert alert-info">An SVG icon should be pasted in the textarea</div>
+<div class="form-group" data-controller="svg-icon-upload">
+  <div nav-link-id="<%= form.object[:id] %>">
```
Comment (rhymes) [inline app/views/admin/navigation_links/_form.html.erb:12]: this should probably be a `data-` attribute

**Review by rhymes — COMMENTED** [review]
Thank you @joaoGabriel55! Works well, I added some technical notes about simplifying the code by taking advantage of Stimulus's own features

File: app/javascript/admin/controllers/svg_icon_upload_controller.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    );
+
+    this.clearInvalidIconTypeMessage(currentNavigationLinkModal);
+
+    const svgIconContent = currentNavigationLinkModal.querySelector(
+      '#svg-icon-content',
+    );
+
+    const icon = event.target.files[0];
+    if (icon.type !== 'image/svg+xml') {
```
[the remark this answers is not in the document]
Comment (joaoGabriel55) [reply app/javascript/admin/controllers/svg_icon_upload_controller.js:16]: I guess is good do this defensive programming to check the file type in the controller. No one knows what the final user can do. So, I think keep the two validations would be good. What do you think?

File: app/javascript/admin/controllers/svg_icon_upload_controller.js (hunk 2 above)
Comment (joaoGabriel55) [reply app/javascript/admin/controllers/svg_icon_upload_controller.js:42]: Okay! Thanks a lot for the tip

File: app/javascript/admin/controllers/svg_icon_upload_controller.js (hunk 3 above)
Comment (joaoGabriel55) [reply app/javascript/admin/controllers/svg_icon_upload_controller.js:13]: Okay! Thanks a lot for the tip

File: app/javascript/admin/controllers/svg_icon_upload_controller.js (hunk 4 above)
Comment (joaoGabriel55) [reply app/javascript/admin/controllers/svg_icon_upload_controller.js:50]: Okay! Thanks a lot for the tip

File: app/javascript/admin/controllers/svg_icon_upload_controller.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (rhymes) [reply app/javascript/admin/controllers/svg_icon_upload_controller.js]: Sure! Let's keep them both :)

File: app/views/admin/navigation_links/_form.html.erb [hunk 8]
```diff
@@ -9,19 +9,19 @@
...
-    <div id="svg-icon-preview" class="<%= form.object[:id].nil? ? "" : "pb-3" %>" style="width: 24px;">
+    <div data-svg-icon-upload-target="svgIconPreview" class="<%= form.object[:id].nil? ? "" : "pb-3" %>" style="width: 48px;">
       <%= form.object[:icon]&.html_safe %>
     </div>
-    <%= form.hidden_field :icon, id: "svg-icon-content" %>
+    <%= form.hidden_field :icon, "data-svg-icon-upload-target": "svgIconContent" %>
 
     <input
-      type="file" accept="image/svg" class="form-control"
+      type="file" accept="image/svg, image/svg+xml" class="form-control"
```
Comment (rhymes) [inline app/views/admin/navigation_links/_form.html.erb:20]: Not sure all browser support the `, `, according to the specs there should be no space between the types. Given that, `image/svg` is not needed as it doesn't exist :), SVG has only `image/svg+xml` as media type: https://www.w3.org/TR/SVG11/mimereg.html

File: app/javascript/admin/controllers/svg_icon_upload_controller.js [hunk 9]
```diff
@@ -1,29 +1,26 @@
...
-      const navigationLinkId = currentNavigationLinkModal.getAttribute(
-        'nav-link-id',
-      );
-
-      const ableToRemoveSvgIconContent =
-        svgIconContent.value && !navigationLinkId;
-      if (ableToRemoveSvgIconContent) {
-        svgIconContent.value = null;
-        this.setSvgIconPreview(currentNavigationLinkModal, null);
+    if (icon.type !== 'image/svg+xml' && icon.type !== 'image/svg') {
```
Comment (rhymes) [inline app/javascript/admin/controllers/svg_icon_upload_controller.js:15]: See my note about not needing `image/svg` down here

**Review by rhymes — COMMENTED** [review]
Great, we're almost there :) Thank you again @joaoGabriel55

File: app/views/admin/navigation_links/_form.html.erb (hunk 8 above)
Comment (joaoGabriel55) [reply app/views/admin/navigation_links/_form.html.erb:20]: Oh, okay understood.

**Review by rhymes — APPROVED** [review]
LGTM!

File: app/javascript/admin/controllers/svg_icon_upload_controller.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        'alert',
+        'alert-danger',
+      );
+      this.svgIconMessageValidateTarget.innerHTML = null;
+    }
+  }
+
+  invalidIconTypeMessage(type) {
+    this.svgIconMessageValidateTarget.classList.add('alert', 'alert-danger');
+    this.svgIconMessageValidateTarget.innerHTML = `'${type}' is a invalid Icon type`;
```
Comment (aitchiss) [inline app/javascript/admin/controllers/svg_icon_upload_controller.js:58]:
```suggestion
    this.svgIconMessageValidateTarget.innerHTML = `'${type}' is an invalid Icon type`;
```

refactor/blocking - Just a small typo here

File: app/views/admin/navigation_links/_form.html.erb [hunk 11]
```diff
@@ -8,10 +8,21 @@
...
+<div class="form-group" data-controller="svg-icon-upload">
+  <div data-svg-icon-upload-target="navId" nav-link-id="<%= form.object[:id] %>">
+    <%= form.label :icon %>
+    <div data-svg-icon-upload-target="svgIconPreview" class="<%= form.object[:id].nil? ? "" : "pb-3" %>" style="width: 48px;">
+      <%= form.object[:icon]&.html_safe %>
+    </div>
+    <%= form.hidden_field :icon, "data-svg-icon-upload-target": "svgIconContent" %>
+
+    <input
+      type="file" accept="image/svg+xml" class="form-control"
```
Comment (aitchiss) [inline app/views/admin/navigation_links/_form.html.erb:20]:
```suggestion
      type="file" accept="image/svg+xml" class="form-control-file"
```

Refactor/blocking - I think the normal input styling looks a little funky with the file upload option - on Chrome and Safari in particular we have spacing above the button but not below, and the border maybe isn't suitable for this input type. 

I suggest we change to the styling for `form-control-file` which gives us something more specific to this input type.

Before on Chrome:

screenshot showing button with spacing at top but not bottom

After:

screenshot showing a more styled button without the bordered input

**Review by aitchiss — CHANGES_REQUESTED** [review]
This functionality is looking great! I just have a couple of very small change requests for the UI, but everything is working great as far as I can see - nice work 👍

File: app/javascript/admin/controllers/svg_icon_upload_controller.js (hunk 10 above)
Comment (joaoGabriel55) [reply app/javascript/admin/controllers/svg_icon_upload_controller.js:58]: Thanks!

File: app/views/admin/navigation_links/_form.html.erb (hunk 11 above)
Comment (joaoGabriel55) [reply app/views/admin/navigation_links/_form.html.erb]: Done @aitchiss!

**Review by aitchiss — APPROVED** [review]
Thanks for making those changes @joaoGabriel55 - looks good to me 🚀

**Pull request merged**